crates/ruff/src/rules/flake8_pyi/rules/duplicate_types_in_union.rs

```diff
@@ -0,0 +1,72 @@
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::comparable::ComparableExpr;
+use ruff_python_ast::types::Range;
+use rustc_hash::FxHashSet;
+use rustpython_parser::ast::{Expr, ExprKind, Operator};
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+#[violation]
+pub struct DuplicateTypesInUnion {
+    pub duplicate_name: String,
+}
+
+impl AlwaysAutofixableViolation for DuplicateTypesInUnion {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Duplicate {} in union", self.duplicate_name)
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Remove latter {} from union", self.duplicate_name)
+    }
+}
+
+///PYI016
+pub fn duplicate_types_in_union<'a>(
+    mut seen_nodes: FxHashSet<ComparableExpr<'a>>,
```

Comment:
You are passing and returning `FxHashSet` by value, which means it's (theoretically) copied each time you pass and return it. Consider instead taking a `&mut FxHashSet<ComparableExpr<'a>>` and returning nothing. This will eliminate the copies.

Reply:
I had some issues with the recursive calls when I did this thanks to the borrow checker. But I'll give it another crack now that I've got lifetime annotations here.